tools/envoy_headersplit/headersplit.py

```diff
@@ -0,0 +1,394 @@
+    # otherwise we get definitions for every file included
+    # when clang parsed the input file (i.e. if we don't limit descendant loction,
+    # it will check definitions from included headers and get class defns like std::string)
+    if descendant.location.file is None:
+      continue
+    if descendant.location.file.name != cursor.displayname:
+      continue
+    cursors.append(descendant)
+  return cursors
+
+
+def class_definitions(cursor: Cursor) -> List[Cursor]:
+  """
+    extracts all class definitions in the file pointed by cursor. (typical mocks.h)
+
+    Args:
+        cursor: cursor of parsing result of target source code by libclang
+
+    Returns:
+        a list of cursor, each pointing to a class definition.
+    """
+  cursors = cursors_in_same_file(cursor)
+  class_cursors = []
+  for descendant in cursors:
+    # check if descendant is pointing to a class delaration block.
+    if descendant.kind != CursorKind.CLASS_DECL:
+      continue
+    if not descendant.is_definition():
+      continue
+    # check if this class is directly enclosed by a namespace.
+    if descendant.semantic_parent.kind != CursorKind.NAMESPACE:
+      continue
+    class_cursors.append(descendant)
+  return class_cursors
+
+
+def class_implementations(cursor: Cursor) -> List[Cursor]:
+  """
+    extracts all class implementation in the file pointed by cursor. (typical mocks.cc)
+
+    Args:
+        cursor: cursor of parsing result of target source code by libclang
+
+    Returns:
+        a list of cursor, each pointing to a class implementation.
+    """
+  cursors = cursors_in_same_file(cursor)
+  impl_cursors = []
+  for descendant in cursors:
+    if descendant.kind == CursorKind.NAMESPACE:
+      continue
+    # check if descendant is pointing to a class method
+    if descendant.semantic_parent is not None and descendant.semantic_parent.kind == CursorKind.CLASS_DECL:
+      impl_cursors.append(descendant)
+  return impl_cursors
+
+
+def extract_definition(cursor: Cursor, classnames: List[str]) -> Tuple[str, str, List[str]]:
+  """
+    extracts class definition source code pointed by the cursor parameter.
+    and find dependent mock classes by naming look up.
+
+    Args:
+        cursor: libclang cursor pointing to the target mock class definition.
+        classnames: all mock class names defined in the definition header that needs to be
+            divided, used to parse class dependencies.
+    Returns:
+        class_name: a string representing the mock class name.
+        class_defn: a string contatins the whole class definition body.
+        deps: a set of string contatins all dependent classes for the return class.
+
+    Note:
+        It can not detect and resolve forward declaration and cyclic dependency. Need to address
+        manually.
+    """
+  filename = cursor.location.file.name
+  with open(filename, 'r') as source_file:
+    contents = source_file.read()
```

Comment:
Can you add a `read_file_contents(path)` function? We can factor all these `with, read()` blocks out, and potentially add memoization there later (not needed in this PR).